atc/exec/task_config_source.go

```diff
@@ -133,40 +132,51 @@ func (configSource FileConfigSource) Warnings() []string {
 	return []string{}
 }
 
-// MergedConfigSource is used to join two config sources together.
-type MergedConfigSource struct {
-	A             TaskConfigSource
-	B             TaskConfigSource
-	MergeWarnings []string
+// OverrideParamsConfigSource is used to override params in a config source
+type OverrideParamsConfigSource struct {
+	ConfigSource TaskConfigSource
+	Params       atc.Params
 }
 
-// FetchConfig fetches both config sources, and merges the second config source
-// into the first. This allows the user to set params required by a task loaded
+// FetchConfig overrides parameters, allowing the user to set params required by a task loaded
 // from a file by providing them in static configuration.
-func (configSource *MergedConfigSource) FetchConfig(logger lager.Logger, source *worker.ArtifactRepository) (atc.TaskConfig, error) {
-	aConfig, err := configSource.A.FetchConfig(logger, source)
+func (configSource OverrideParamsConfigSource) FetchConfig(logger lager.Logger, source *worker.ArtifactRepository) (atc.TaskConfig, error) {
+	taskConfig, err := configSource.ConfigSource.FetchConfig(logger, source)
 	if err != nil {
 		return atc.TaskConfig{}, err
 	}
 
-	bConfig, err := configSource.B.FetchConfig(logger, source)
-	if err != nil {
-		return atc.TaskConfig{}, err
+	if taskConfig.Params == nil {
+		taskConfig.Params = map[string]string{}
 	}
 
-	mergedConfig, warnings, err := aConfig.Merge(bConfig)
-	configSource.MergeWarnings = warnings
+	for key, val := range configSource.Params {
+		switch v := val.(type) {
+		case string:
+			taskConfig.Params[key] = v
+		case float64:
+			if math.Floor(v) == v {
+				taskConfig.Params[key] = strconv.FormatInt(int64(v), 10)
+			} else {
+				taskConfig.Params[key] = strconv.FormatFloat(v, 'f', -1, 64)
+			}
+		default:
+			bs, err := json.Marshal(val)
+			if err != nil {
+				return atc.TaskConfig{}, err
+			}
+			taskConfig.Params[key] = string(bs)
+		}
+	}
 
-	return mergedConfig, err
+	return taskConfig, nil
 }
 
-func (configSource *MergedConfigSource) Warnings() []string {
-	warnings := []string{}
-	warnings = append(warnings, configSource.A.Warnings()...)
-	warnings = append(warnings, configSource.B.Warnings()...)
-	warnings = append(warnings, configSource.MergeWarnings...)
-
-	return warnings
+func (configSource OverrideParamsConfigSource) Warnings() []string {
+	if len(configSource.Params) > 0 {
+		return []string{fmt.Sprintf("overriding task parameters via 'params' is deprecated. use 'vars' instead")}
```

Comment:
:thinking: I don't think `vars:` really replaces `params:`, or at least we haven't decided to deprecate it at this point. Params will still be useful for things like tasks that have default values, plus allowing arbitrary env config (https://github.com/concourse/concourse/issues/2560):

**task.yml**

```
---
platform: linux

# ...

params:
  FOO: bar
```

**pipeline.yml**

```
task: foo
file: thing/task.yml
params: {FOO: buzz}
```

With `((vars))` it's all-or-nothing - defaults can't be specified.

I think it's safer at this point to introduce `vars:` and see how they're used for now. The concourse/rfcs#7 (future RFC) will be a good opportunity to reassess everything.

Reply:
ok, makes sense. let's keep them both